apps/src/netsim/NetSimVisualization.js

```diff
@@ -0,0 +1,475 @@
+    if (entity.isDead()) {
+      entity.getRoot().remove();
+      return false;
+    }
+    return true;
+  });
+};
+
+/**
+ * Render: Let all vizentities "redraw" (or in our case, touch the DOM)
+ */
+NetSimVisualization.prototype.render = function () {
+  this.entities_.forEach(function (entity) {
+    entity.render();
+  });
+};
+
+/**
+ * Called whenever the connection notifies us that we've connected to,
+ * or disconnected from, a shard.
+ * @param {?NetSimShard} newShard - null if disconnected.
+ * @param {?NetSimLocalClientNode} localNode - null if disconnected
+ * @private
+ */
+NetSimVisualization.prototype.onShardChange_= function (newShard, localNode) {
+  this.setShard(newShard);
+  this.setLocalNode(localNode);
+};
+
+/**
+ * Change the shard this visualization will source its data from.
+ * Re-attaches table change listeners for all the tables we need to monitor.
+ * @param {?NetSimShard} newShard - null if disconnected
+ */
+NetSimVisualization.prototype.setShard = function (newShard) {
+  if (this.nodeTableChangeKey !== undefined) {
+    this.shard_.nodeTable.tableChange.unregister(this.nodeTableChangeKey);
+    this.nodeTableChangeKey = undefined;
+  }
+
+  if (this.wireTableChangeKey !== undefined) {
+    this.shard_.wireTable.tableChange.unregister(this.wireTableChangeKey);
+    this.wireTableChangeKey = undefined;
+  }
+
+  this.shard_ = newShard;
+  if (!this.shard_) {
+    return;
+  }
+
+  this.nodeTableChangeKey = this.shard_.nodeTable.tableChange.register(
+      this.onNodeTableChange_.bind(this));
+
+  this.wireTableChangeKey = this.shard_.wireTable.tableChange.register(
+      this.onWireTableChange_.bind(this));
+};
+
+/**
+ * Change which node we consider the 'local node' in the visualization.
+ * We go through a special creation process for this node, so that it
+ * looks and behaves differently.
+ * @param {?NetSimLocalClientNode} newLocalNode - null if disconnected
+ */
+NetSimVisualization.prototype.setLocalNode = function (newLocalNode) {
+  if (newLocalNode) {
+    if (this.localNode) {
+      this.localNode.configureFrom(newLocalNode);
+    } else {
+      this.localNode = new NetSimVizNode(newLocalNode);
+      this.entities_.push(this.localNode);
+      this.svgRoot_.find('#background_group').append(this.localNode.getRoot());
+    }
+  } else {
+    this.localNode.kill();
+  }
+  this.pullElementsToForeground();
+};
+
+/**
+ * Find a particular VizEntity in the visualization, by type and ID.
+ * @param {function} entityType - constructor of entity we're looking for
+ * @param {number} entityID - ID, with corresponds to NetSimEntity.entityID
+ * @returns {NetSimVizEntity} or null if not found
+ */
+NetSimVisualization.prototype.getEntityByID = function (entityType, entityID) {
+  return this.entities_.reduce(function (prev, cur) {
+    if (prev) {
+      return prev;
+    } else if (cur instanceof entityType && cur.id === entityID) {
+      return cur;
+    }
+    return null;
+  }, null);
```

Comment:
Is this just a find? Should we use lodash instead? Doing in a reduce this way seems a little funky.